packages/server-wallet/src/wallet/wallet.ts

```diff
@@ -321,6 +331,137 @@ export class Wallet extends EventEmitter<WalletEvents> {
+  }
+
+  /**
+   * Emits events, sends messages and requests transactions based on the output of the engine.
+   * @param output
+   */
+  private async handleEngineOutput(
+    output: MultipleChannelOutput | SingleChannelOutput
+  ): Promise<void> {
+    this.emitObjectiveEvents(output);
+    await this._messageService.send(getMessages(output));
+    await this.waitForTransactions(this._chainService.handleChainRequests(output.chainRequests));
+  }
+
+  /**
+   * Creates a listener that will be subscribed to the chain service.
+   * It creates a listener for a specific channel id.
+   * @param channelIdToListenFor The channel Id the listener is for
+   * @returns A chain event subscriber that can be passed into the chain service.
+   */
+  private createChainListener(channelIdToListenFor: string): ChainEventSubscriberInterface {
+    return {
+      holdingUpdated: async ({channelId, amount, assetHolderAddress}) => {
+        if (channelId !== channelIdToListenFor) return;
+
+        this._engine.logger.trace({channelId, amount}, 'holdingUpdated');
+        try {
+          await this._engine.store.updateFunding(channelId, amount, assetHolderAddress);
+          const result = await this._engine.crank([channelId]);
+          await this.handleEngineOutput(result);
+        } catch (err) {
+          this._engine.logger.error(err, 'holdingUpdated error');
+          throw err;
+        }
+      },
+      assetOutcomeUpdated: async ({channelId, assetHolderAddress, externalPayouts}) => {
+        try {
+          if (channelId !== channelIdToListenFor) return;
+
+          this._engine.logger.trace(
+            {channelId, assetHolderAddress, externalPayouts},
+            'assetOutcomeUpdated'
+          );
+          const transferredOut = externalPayouts.map(ai => ({
+            toAddress: makeDestination(ai.destination),
+            amount: ai.amount as Uint256,
+          }));
+
+          await this._engine.store.updateTransferredOut(
+            channelId,
+            assetHolderAddress,
+            transferredOut
+          );
+          const result = await this._engine.crank([channelId]);
+          await this.handleEngineOutput(result);
+        } catch (err) {
+          this._engine.logger.error(err, 'assetOutcomeUpdated error');
+          throw err;
+        }
+      },
+      challengeRegistered: async ({channelId, finalizesAt: finalizedAt, challengeStates}) => {
+        try {
+          if (channelId !== channelIdToListenFor) return;
+          this._engine.logger.trace(
+            {channelId, finalizedAt, challengeStates},
+            'challengeRegistered'
+          );
+          await this._engine.store.insertAdjudicatorStatus(channelId, finalizedAt, challengeStates);
+          const result = await this._engine.crank([channelId]);
+          await this.handleEngineOutput(result);
+        } catch (err) {
+          this._engine.logger.error(err, 'challengeRegistered error');
+          throw err;
+        }
+      },
```

Comment:
You're right that these all follow the same pattern. Is there a way to encapsulate that pattern in code?

as an idea:
```
function createHandler(name, inserter) {
  return async (data) => {
    if (data.channelId !== channelIdToListenFor) return;
    this._engine.logger.trace(data, name);
    await inserter(data);
    await this._engine.crank([channelId]);
    await this.handleEngineOutput(result);
  }
}
```

Reply:
Yup. I was hoping to do something like that but never got around it

Reply:
I've updated this in https://github.com/statechannels/statechannels/pull/3564/commits/2c21498ab50446a9880fb66d48fe9db8df374d5b